plugins/ntn_airswap/ntn_airswap.go

```diff
@@ -0,0 +1,445 @@
+	chNTNTransferEvent := make(chan *erc20.Erc20Transfer)
+	subNTNTransferEvent, err := e.ntnContract.WatchTransfer(new(bind.WatchOpts), chNTNTransferEvent, nil, nil)
+	if err != nil {
+		e.logger.Error("cannot watch NTN transfer event", "error", err)
+		return err
+	}
+	// subscribe on-chain transfer event of USDC.
+	chUSDCTransferEvent := make(chan *erc20.Erc20Transfer)
+	subUSDCTransferEvent, err := e.usdcContract.WatchTransfer(new(bind.WatchOpts), chUSDCTransferEvent, nil, nil)
+	if err != nil {
+		e.logger.Error("cannot watch USDC transfer event", "error", err)
+		return err
+	}
+	// subscribe on-chain swap event of airsapERC20.
+	chSwapEvent := make(chan *swaperc20.Swaperc20SwapERC20)
+	subSwapEvent, err := e.swapContract.WatchSwapERC20(new(bind.WatchOpts), chSwapEvent, nil, nil)
+	if err != nil {
+		e.logger.Error("cannot watch swap event", "error", err)
+		return err
+	}
+	e.chUSDCTransferEvent = chUSDCTransferEvent
+	e.subUSDCTransferEvent = subUSDCTransferEvent
+	e.chNTNTransferEvent = chNTNTransferEvent
+	e.subNTNTransferEvent = subNTNTransferEvent
+	e.chSwapEvent = chSwapEvent
+	e.subSwapEvent = subSwapEvent
+	return nil
+}
+
+func (e *AirswapClient) StartWatcher() {
+	for {
+		select {
+		case <-e.doneCh:
+			e.ticker.Stop()
+			e.logger.Info("air-swap events watcher stopped")
+			return
+		case err := <-e.subSwapEvent.Err():
+			if err != nil {
+				e.logger.Info("subscription error of swap event", "error", err)
+				e.handleConnectivityError()
+				e.subSwapEvent.Unsubscribe()
+			}
+		case err := <-e.subUSDCTransferEvent.Err():
+			if err != nil {
+				e.logger.Info("subscription error of usdc transfer event", "error", err)
+				e.handleConnectivityError()
+				e.subUSDCTransferEvent.Unsubscribe()
+			}
+		case err := <-e.subSwapEvent.Err():
+			if err != nil {
+				e.logger.Info("subscription error of swap event", "error", err)
+				e.handleConnectivityError()
+				e.subSwapEvent.Unsubscribe()
+			}
+		case ntnTransferEvent := <-e.chNTNTransferEvent:
+			e.logger.Debug("receiving ntn transfer event", "event", ntnTransferEvent)
+			transfer := &SwapEvent{ntnTransfer: ntnTransferEvent}
+			e.handleEvent(ntnTransferEvent.Raw.TxHash, transfer)
+		case usdcTransferEvent := <-e.chUSDCTransferEvent:
+			e.logger.Debug("receiving usdc transfer event", "event", usdcTransferEvent)
+			transfer := &SwapEvent{usdcTransfer: usdcTransferEvent}
+			e.handleEvent(usdcTransferEvent.Raw.TxHash, transfer)
+		case airSwapEvent := <-e.chSwapEvent:
+			e.logger.Debug("receiving air swap event", "event", airSwapEvent)
+			swap := &SwapEvent{airSwap: airSwapEvent}
+			e.handleEvent(airSwapEvent.Raw.TxHash, swap)
+		case <-e.ticker.C:
+			e.checkHealth()
+		}
+	}
+}
+
+func (e *AirswapClient) handleEvent(txnHash ecommon.Hash, swapEvent *SwapEvent) {
+
+	// if event was cached, fill the sub events, and check if we can push it to the aggregator.
+	if cached, ok := e.swapEventCache.Get(txnHash); ok {
+		swap := cached.(*SwapEvent)
+		swap.fill(swapEvent)
```

Comment:
Addressed a bug here, as the swap() can be called by smart contract, we should take the swap event pairing more carefully. Todo: fix this bug.

Reply:
Resolved in 45c593da30c7135b30fe13c60fe387c8ec3b725b and 0766faf34347eb44864acb4e495f9f1a25ccb172